## Importing Libraries

In [23]:
#Libraries Import
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import LabelEncoder
from scipy import sparse
from imblearn.over_sampling import SMOTE

## Loading Dataset

In [24]:
#Loading the dataset
file_path = '/content/sample_data/hotel_bookings.csv'
hotel_data = pd.read_csv(file_path)

#Displaying the first few rows and general info
hotel_data.head(), hotel_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

(          hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
 0  Resort Hotel            0        342               2015               July   
 1  Resort Hotel            0        737               2015               July   
 2  Resort Hotel            0          7               2015               July   
 3  Resort Hotel            0         13               2015               July   
 4  Resort Hotel            0         14               2015               July   
 
    arrival_date_week_number  arrival_date_day_of_month  \
 0                        27                          1   
 1                        27                          1   
 2                        27                          1   
 3                        27                          1   
 4                        27                          1   
 
    stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
 0                        0                     0       2  ...    No D

## Handle Missing Values

In [25]:
#Replacing missing values with '0' in Children; Country, Agent, Company missing values with 'Unknown'
hotel_data['children'] = hotel_data['children'].fillna(0)
hotel_data['country'] = hotel_data['country'].fillna('Unknown')

hotel_data['agent'] = hotel_data['agent'].fillna('Unknown')
hotel_data['company'] = hotel_data['company'].fillna('Unknown')

#Checking if any more remaining missing values
hotel_data.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
agent                             0
company                           0
days_in_waiting_list              0
customer_type                     0
adr                               0
required_car_parking_spaces       0
total_of_special_requests         0
reservation_status                0
reservation_status_date           0
dtype: int64

## Feature Engineering

In [26]:
#Creating new features, such as the total number of nights, and combine date columns to a single datetime feature.

#'total_nights' feature by combining stays in weekend and week nights
hotel_data['total_nights'] = hotel_data['stays_in_weekend_nights'] + hotel_data['stays_in_week_nights']

#A combined 'date' feature using year, month, and day
hotel_data['date'] = pd.to_datetime(hotel_data[['arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month']].astype(str).agg('-'.join, axis=1))

#Removing the individual date columns now that we have the combined 'date'
hotel_data.drop(columns=['arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month'], inplace=True)

#Checking the new features
hotel_data[['total_nights', 'date']].head()

total_nights       date
0             0 2015-07-01
1             0 2015-07-01
2             1 2015-07-01
3             1 2015-07-01
4             2 2015-07-01

In [27]:
#Separate features (X) and target (y)
X = hotel_data.drop(columns=['is_canceled'])
y = hotel_data['is_canceled']

#Spliting the data into training and testing sets (80% train, 20% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#Check the split dimensions
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((95512, 30), (23878, 30), (95512,), (23878,))

## Encoding Categorical Features

In [28]:
#One-Hot Encoding for multi-category columns
categorical_cols = ['hotel', 'meal', 'market_segment', 'distribution_channel', 'reserved_room_type', 'assigned_room_type', 'customer_type']
hotel_data_encoded = pd.get_dummies(hotel_data, columns=categorical_cols, drop_first=True)

#Label Encoding for binary categorical features
label_encoders = ['deposit_type', 'is_repeated_guest']
le = LabelEncoder()
for col in label_encoders:
    hotel_data_encoded[col] = le.fit_transform(hotel_data_encoded[col])

#Checking the updated dataset
hotel_data_encoded.head()


is_canceled  lead_time  arrival_date_week_number  stays_in_weekend_nights  \
0            0        342                        27                        0   
1            0        737                        27                        0   
2            0          7                        27                        0   
3            0         13                        27                        0   
4            0         14                        27                        0   

   stays_in_week_nights  adults  children  babies country  is_repeated_guest  \
0                     0       2       0.0       0     PRT                  0   
1                     0       2       0.0       0     PRT                  0   
2                     1       1       0.0       0     GBR                  0   
3                     1       1       0.0       0     GBR                  0   
4                     2       2       0.0       0     GBR                  0   

   ...  assigned_room_type_F  assigned_room_type_G  assigned_room_type_H  \
0  ...                 False                 False                 False   
1  ...                 False                 False                 False   
2  ...                 False                 False                 False   
3  ...                 False                 False                 False   
4  ...                 False                 False                 False   

   assigned_room_type_I assigned_room_type_K assigned_room_type_L  \
0                 False                False                False   
1                 False                False                False   
2                 False                False                False   
3                 False                False                False   
4                 False                False                False   

   assigned_room_type_P  customer_type_Group  customer_type_Transient  \
0                 False                False                     True   
1                 False                False                     True   
2                 False                False                     True   
3                 False                False                     True   
4                 False                False                     True   

   customer_type_Transient-Party  
0                          False  
1                          False  
2                          False  
3                          False  
4                          False  

[5 rows x 63 columns]

## Normalization

In [29]:
#List of categorical columns
categorical_cols = ['hotel', 'meal', 'market_segment', 'distribution_channel', 'reserved_room_type', 'assigned_room_type', 'customer_type', 'country']
hotel_data_encoded = pd.get_dummies(hotel_data, columns=categorical_cols, drop_first=True)

#List of numerical features to scale
numerical_cols = ['lead_time', 'total_nights', 'adults', 'children', 'previous_cancellations', 'previous_bookings_not_canceled', 'booking_changes', 'days_in_waiting_list', 'adr', 'required_car_parking_spaces', 'total_of_special_requests']

X = hotel_data_encoded.drop(columns=['is_canceled'])
y = hotel_data_encoded['is_canceled']

## Pipeline setup

In [30]:
#Preprocessing pipeline for numerical data
numerical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),  # Handle missing values in numerical columns
    ('scaler', StandardScaler())  # Scale numerical columns
])

categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='Unknown')),  # Handle missing values in categorical columns
    ('encoder', OneHotEncoder(handle_unknown='ignore'))  # One-hot encode categorical columns
])

# Combine the two transformers into one preprocessing pipeline
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_transformer, numerical_cols),
        ('cat', categorical_transformer, categorical_cols)
    ]
)

# Fit and transform the training data
X_train_processed = preprocessor.fit_transform(X_train)
X_test_processed = preprocessor.transform(X_test)

# Check the processed data dimensions
X_train_processed.shape, X_test_processed.shape

((95512, 231), (23878, 231))

## Class Imbalance Handling

In [31]:
#Applying SMOTE on numerical features
smote = SMOTE(random_state=42)
X_resampled, y_resampled = smote.fit_resample(X_train_processed, y_train)

#Check the class distribution after SMOTE
y_resampled.value_counts()

is_canceled
1    60259
0    60259
Name: count, dtype: int64

## Saving Test and Train Splits

In [32]:
#Saving the processed train-test splits as CSV files
X_train_processed_df = pd.DataFrame(X_train_processed)
X_test_processed_df = pd.DataFrame(X_test_processed)
y_train_df = pd.DataFrame(y_train)
y_test_df = pd.DataFrame(y_test)

if sparse.issparse(X_train_processed):
    X_train_processed_df = pd.DataFrame(X_train_processed.toarray())
    X_test_processed_df  = pd.DataFrame(X_test_processed.toarray())
else:
    X_train_processed_df = pd.DataFrame(X_train_processed)
    X_test_processed_df  = pd.DataFrame(X_test_processed)

y_train_df = pd.DataFrame(y_train)
y_test_df  = pd.DataFrame(y_test)

#Save to CSV
X_train_processed_df.to_csv('X_train_processed.csv', index=False)
X_test_processed_df.to_csv('X_test_processed.csv', index=False)
y_train_df.to_csv('y_train.csv', index=False)
y_test_df.to_csv('y_test.csv', index=False)
#Verify that they are saved correctly
print("Files saved successfully!")


Files saved successfully!


In [33]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [34]:
X_train = pd.read_csv("X_train_processed.csv")
print(X_train.dtypes.head())
print(X_train.shape)

0    float64
1    float64
2    float64
3    float64
4    float64
dtype: object
(95512, 231)


Random Forest

In [35]:
import pandas as pd
import numpy as np

from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    accuracy_score,
    roc_auc_score,
    confusion_matrix,
    classification_report
)

from imblearn.over_sampling import SMOTE
from sklearn.model_selection import GridSearchCV


In [36]:
X_train = pd.read_csv("X_train_processed.csv")
X_test  = pd.read_csv("X_test_processed.csv")
y_train = pd.read_csv("y_train.csv").squeeze()
y_test  = pd.read_csv("y_test.csv").squeeze()

print("X_train:", X_train.shape)
print("X_test: ", X_test.shape)
print("y_train:", y_train.shape)
print("y_test: ", y_test.shape)


X_train: (95512, 231)
X_test:  (23878, 231)
y_train: (95512,)
y_test:  (23878,)


In [37]:
smote = SMOTE(random_state=42)
X_train_balanced, y_train_balanced = smote.fit_resample(X_train, y_train)

print("Before SMOTE:\n", y_train.value_counts())
print("\nAfter SMOTE:\n", y_train_balanced.value_counts())


Before SMOTE:
 is_canceled
0    60259
1    35253
Name: count, dtype: int64

After SMOTE:
 is_canceled
1    60259
0    60259
Name: count, dtype: int64


Train a BASELINE Random Forest Model

In [38]:
rf = RandomForestClassifier(
    n_estimators=200,
    random_state=42,
    n_jobs=-1
)

rf.fit(X_train_balanced, y_train_balanced)

y_pred = rf.predict(X_test)
y_proba = rf.predict_proba(X_test)[:, 1]

print("BASELINE RANDOM FOREST RESULTS")
print("Accuracy:", accuracy_score(y_test, y_pred))
print("ROC-AUC :", roc_auc_score(y_test, y_proba))

print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))


BASELINE RANDOM FOREST RESULTS
Accuracy: 0.8818996565876539
ROC-AUC : 0.9496238960004822

Confusion Matrix:
 [[13595  1312]
 [ 1508  7463]]

Classification Report:
               precision    recall  f1-score   support

           0       0.90      0.91      0.91     14907
           1       0.85      0.83      0.84      8971

    accuracy                           0.88     23878
   macro avg       0.88      0.87      0.87     23878
weighted avg       0.88      0.88      0.88     23878



5️⃣ Hyperparameter Tuning — Random Forest (GridSearchCV)

**Evaluate the BEST Random Forest Model**

In [21]:
# Quick sanity check - runs in ~1-2 minutes
quick_params = {
    "n_estimators": [100, 200],
    "max_depth": [10, 20],
    "max_features": ["sqrt", "log2"]
}

rand_search_rf = RandomizedSearchCV(
    RandomForestClassifier(random_state=42),
    quick_params,
    n_iter=4,  # Just 4 iterations = 12 fits total
    scoring="roc_auc",
    cv=3,
    n_jobs=-1,
    verbose=2,
    random_state=42
)

rand_search_rf.fit(X_train_balanced, y_train_balanced)
print("Best parameters:", rand_search_rf.best_params_)
print("Best CV ROC-AUC:", rand_search_rf.best_score_)

Fitting 3 folds for each of 4 candidates, totalling 12 fits


/usr/local/lib/python3.12/dist-packages/joblib/externals/loky/process_executor.py:782: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
  warnings.warn(


Best parameters: {'n_estimators': 200, 'max_features': 'sqrt', 'max_depth': 20}
Best CV ROC-AUC: 0.9473487431359674


In [39]:
param_dist = {
    "n_estimators": [100, 200, 300],
    "max_depth": [10, 15, 20, 30],
    "min_samples_split": [2, 5, 10],
    "min_samples_leaf": [1, 2, 4],
    "max_features": ["sqrt", "log2"]
}

rand_search_rf = RandomizedSearchCV(
    RandomForestClassifier(random_state=42),
    param_dist,
    n_iter=10,  # 30 fits total
    scoring="roc_auc",
    cv=3,
    n_jobs=-1,
    verbose=2,
    random_state=42
)

rand_search_rf.fit(X_train_balanced, y_train_balanced)
print("Best parameters:", rand_search_rf.best_params_)
print("Best CV ROC-AUC:", rand_search_rf.best_score_)

Fitting 3 folds for each of 10 candidates, totalling 30 fits
Best parameters: {'n_estimators': 100, 'min_samples_split': 5, 'min_samples_leaf': 1, 'max_features': 'sqrt', 'max_depth': 20}
Best CV ROC-AUC: 0.9456950120714976


In [43]:
best_rf = rand_search_rf.best_estimator_

y_pred_best = best_rf.predict(X_test)
y_proba_best = best_rf.predict_proba(X_test)[:, 1]

print("\nBEST RANDOM FOREST RESULTS ON TEST SET")
print("Accuracy:", accuracy_score(y_test, y_pred_best))
print("ROC-AUC:", roc_auc_score(y_test, y_proba_best))

print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred_best))
print("\nClassification Report:\n", classification_report(y_test, y_pred_best))




BEST RANDOM FOREST RESULTS ON TEST SET
Accuracy: 0.8621325069101264
ROC-AUC: 0.9391195762630326

Confusion Matrix:
 [[13184  1723]
 [ 1569  7402]]

Classification Report:
               precision    recall  f1-score   support

           0       0.89      0.88      0.89     14907
           1       0.81      0.83      0.82      8971

    accuracy                           0.86     23878
   macro avg       0.85      0.85      0.85     23878
weighted avg       0.86      0.86      0.86     23878



# OUTPUT:
# Accuracy: 0.862 (86.2%) - Overall correct predictions
# ROC-AUC: 0.939 (93.9%) - Excellent class separation ability
#
# Confusion Matrix:
# [[13184  1723]  → 13,184 correctly predicted non-churners | 1,723 false alarms
#  [ 1569  7402]] → 1,569 missed churners | 7,402 correctly predicted churners
#
# Classification Report:
# Class 0 (Non-Churn): Precision 89%, Recall 88%, F1 89%
# Class 1 (Churn): Precision 81%, Recall 83%, F1 82%
#
# Key Takeaway: Model catches 83% of churners with 81% precision - strong performance
# for targeting retention campaigns.